apps/protoc-gen-openapi/main.go

```diff
@@ -20,23 +20,28 @@ import (
 
 	"github.com/google/gnostic/apps/protoc-gen-openapi/generator"
 	"google.golang.org/protobuf/compiler/protogen"
+	"google.golang.org/protobuf/types/pluginpb"
 )
 
 var flags flag.FlagSet
 
 func main() {
 	conf := generator.Configuration{
-		Version:     flags.String("version", "0.0.1", "version number text, e.g. 1.2.3"),
-		Title:       flags.String("title", "", "name of the API"),
-		Description: flags.String("description", "", "description of the API"),
-		Naming:      flags.String("naming", "json", `naming convention. Use "proto" for passing names directly from the proto files`),
+		Version:       flags.String("version", "0.0.1", "version number text, e.g. 1.2.3"),
+		Title:         flags.String("title", "", "name of the API"),
+		Description:   flags.String("description", "", "description of the API"),
+		Naming:        flags.String("naming", "json", `naming convention. Use "proto" for passing names directly from the proto files`),
+		CircularDepth: flags.Int("depth", 2, `depth of recursion for circular messages`),
```

Comment:
Could you point me to any examples in the tests (or elsewhere) that require this? I'm wondering how someone would know what value to use for this.

Reply:
I can see how that's really not obvious from a user perspective...
I just wanted to make sure, that it is configurable, since I can think of cases for wanting both a higher and a lower value.

Lower: if it's references that is only set in code and not expected to be given as query params.
Higher: if it's used for something that resembles a family tree, where the depth could be really high.

I can't seem to figure out, how to link to the patch, so I'll use links to the file in the pull requests.

I start a map containing visit depths here:
https://github.com/google/gnostic/blob/7aa98fc70ada0b953b2698adad65480615933359/apps/protoc-gen-openapi/generator/openapi-v3.go#L366-L372

The map is being used here:
https://github.com/google/gnostic/blob/7aa98fc70ada0b953b2698adad65480615933359/apps/protoc-gen-openapi/generator/openapi-v3.go#L388-L407

With a check on the CircularDepth var here:
https://github.com/google/gnostic/blob/7aa98fc70ada0b953b2698adad65480615933359/apps/protoc-gen-openapi/generator/openapi-v3.go#L397

I have added recursive messages in the protobuftypes test:
https://github.com/google/gnostic/blob/7aa98fc70ada0b953b2698adad65480615933359/apps/protoc-gen-openapi/examples/tests/protobuftypes/message.proto#L48-L55

and added them to Message (line 71 and 76):
https://github.com/google/gnostic/blob/7aa98fc70ada0b953b2698adad65480615933359/apps/protoc-gen-openapi/examples/tests/protobuftypes/message.proto#L67-L76

And the test .json files has been updated to reflect the recursion, e.g.:
https://github.com/google/gnostic/blob/7aa98fc70ada0b953b2698adad65480615933359/apps/protoc-gen-openapi/examples/tests/protobuftypes/openapi.yaml#L26-L45

Currently, only the default (2) is being tested though.